### DISCUSSION TITLE
On the decision to make a pure-buildkit docker package

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#1874] [General]
I wanted to address the decision to 1) make `universe.dagger.io/docker` a pure buildkit client, instead of a docker engine client; 2) move the traditional docker engine client to an obscure sub-package; 3) encourage the use of the former over the latter.

I know that it is not an intuitive decision: it's definitely a tradeoff. But I think it's the right one when you look at the current state and trajectory of the Docker platform.

In short: *docker engine is now the legacy API, and buildkit is its successor.*

* Most of its production clients are wrapper tools, using it as a shim to containerd. This shim has recently been deprecated in Kubernetes/CRI and will eventually be deprecated by all infrastructure software. This is encouraged by Docker, after all they created containerd in the first place, for this purpose.

* Its most used development clients is docker-compose which could, and I believe one day will, target buildkit instead of docker engine. I believe docker-compose build already does this. The buildkit team intends to become a feature-complete container runtime for development workloads, and I have no doubt they will pull it off (they largely already have)

* Docker build & buildx of course uses buildkit entirely.

I believe this split aligns us with the future of Docker rather than its past. It will both help them realize this transition, and help us make Dagger the best third-party client on the market for new dockerd aka buildkitd :)

**Comment by grouville** [comment]
I really like the idea and totally agree with you. The biggest blocker atm is: https://github.com/dagger/dagger/issues/1844, for which the absence of a `docker.#Client` package will force users to use the `bash.#Run` package with the `docker-cli` binary installed, meaning that people will be free / feel the need for such package pretty soon

Do we want to kind of control this behavior ? Shall @helderco make it his own package that people can download to keep the `docker.#Client` clean ?

P.S: you already answered here 😇 (https://github.com/dagger/dagger/pull/1855#issuecomment-1078742175)

**Comment by helderco** [comment]
Thanks for the context @shykes, makes perfect sense to me!

**Comment by helderco** [comment]
> I really like the subpackage name `legacy` as an alternative to `client`. It conveys the intention more clearly (see #1856) and is more accurate since _technically both packages are clients_ - just not to the same API.

_Originally posted by @shykes in https://github.com/dagger/dagger/pull/1855#issuecomment-1078742175_ (emphasis mine)

Makes sense, I have just one small reservation on `legacy`. The Particubes use-case that @grouville is working on uses `#Load`. My concern is having featured in our docs something that's using a "legacy" API (i.e., `legacy.#Load`), that conflicts with our intention to discourage its use.

It could also be `docker/engine`, with `engine.#Load` and `engine.#Run`. Not to confuse with the _**dagger** engine_ in early Europa, but about the _**docker** engine_. It tracks with the terminology to "run in a local or remote engine". Then, when our own buildkit client is more full featured, in a future version we could rename `engine` to `legacy`.

**Comment by helderco** [comment]
Just to record, we decided on `universe.dagger.io/docker/cli`.
